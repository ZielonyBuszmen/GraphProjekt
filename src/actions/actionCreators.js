import {INCREMENT_LIKES} from '../reducers/likesReducer';
import { BLOCK_GRAPH_INPUTS, CHECK_GRAPH_CONNECTIVITY } from '../reducers/connectivityReducer';
import { CREATE_EMPTY_MATRIX, LISTY_SASIEDZTWA_CHANGED, MACIERZ_SASIEDZTWA_CHANGED } from '../reducers/matrixReducer';
import { CHANGE_GRAPH_SIZE } from '../reducers/graphSizeRecuder';

export function increment(ktory, info) {
  return {
    type: INCREMENT_LIKES,
    ktory,
    info,
    dodatkowe: 'jakies inne dane',
  };
}

/**
 * @param graph - array[][], graf w formie macierzy sąsiedztwa
 * @return {{type, graph: *}}
 */
export function checkGraphConnectivity(graph) {
  return {
    type: CHECK_GRAPH_CONNECTIVITY,
    graph,
  }
}

/**
 * @param col - kolumna
 * @param row - wiersz
 * @param value - wartość
 * @returns {{type: string, col: *, row: *, value: *}}
 */
export function macierzSasiedztwaChanged(row, col, value) {
  return {
    type: MACIERZ_SASIEDZTWA_CHANGED,
    row,
    col,
    value,
  }
}

/**
 * @param size
 * @returns {{type: string, size: *}}
 */
export function createEmptyMatrix(size) {
  return {
    type: CREATE_EMPTY_MATRIX,
    size
  }
}

/**
 * @param lists - array[][], graf w formie listy sąsiedztwa
 * @return {{type, graph: *}}
 */
export function listySasiedztwaChanged(lists) {
  return {
    type: LISTY_SASIEDZTWA_CHANGED,
    lists
  }
}

/**
 * @param block - boolean
 * @returns {{type: string, block: boolean}}
 */
export function blockGraphInputs(block = true) {
  return {
    type: BLOCK_GRAPH_INPUTS,
    block
  }
}

export function graphSizeChanged(size) {
  return {
    type: CHANGE_GRAPH_SIZE,
    size
  }
}