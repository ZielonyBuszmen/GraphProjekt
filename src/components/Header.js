import React from 'react';

import {
  Navbar,
  NavbarToggler,
  NavbarBrand,
  Container } from 'reactstrap';

class Header extends React.Component {
  render() {
    return(
      <div>
        <Navbar light expand="md" style={{'backgroundColor': '#290004'}}>
          <Container>
              <NavbarBrand style={{'color': '#efe7e7'}} href="/"><b>Silna spójność grafów</b></NavbarBrand>
          <NavbarToggler />
          </Container>
        </Navbar>
      </div>
    );
  }
}
export default Header;