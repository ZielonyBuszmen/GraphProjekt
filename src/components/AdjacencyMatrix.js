import React from 'react'
import { connect } from 'react-redux'
import { createEmptyMatrix, macierzSasiedztwaChanged } from '../actions/actionCreators';
import { Input } from 'reactstrap';

export class AdjacencyMatrix extends React.Component {

  constructor(props) {
    super(props);
    props.createEmptyMatrix(props.size);
  }

  createNumberRow = (size) => {
    const result = [];
    result.push(<td></td>);
    for (let i = 0; i < size; i++) {
      result.push(<td style={{textAlign: 'center'}}>{i}</td>)
    }
    return <tr>{result}</tr>
  };


  prepareInputsArray = (size) => {
    const rows = [];
    rows.push(this.createNumberRow(size));
    for (let i = 0; i < size; i++) {
      const elements = [];
      for (let j = 0; j < size; j++) {
        elements.push(this.inputRenderer(j, j, i));
      }
      rows.push(
        <tr key={i}>
          <td className='pr-2'>{i}</td>
          {elements}
        </tr>
      );
    }
    return rows;
  };

  inputRenderer = (key, row, col) => {
    if (this.props.matrix.length === 0) return '';
    const value = this.props.matrix[col][row] || '';
    return (
      <td>
        <Input className='form-control' maxlength={2} style={{width: '50px'}} type="text" key={key} value={value}
               placeholder={0}
               onChange={(e) => this.handleChange(e, col, row)}/>
      </td>
    );
  };

  handleChange = (event, row, col) => {
    const value = parseInt(event.target.value) || 0;
    this.props.macierzSasiedztwaChanged(row, col, value);
  };

  render() {
    const {size} = this.props;
    return (
      <div className='macierzSasiedztwa'>
        <table>
          {this.prepareInputsArray(size)}
        </table>
      </div>
    );
  }
}


export const mapStateToProps = (state) => {
  return {
    matrix: state.matrix,
  };
};
export const mapDispatchToProps = (dispatch) => {
  return {
    macierzSasiedztwaChanged: (row, col, value) => dispatch(macierzSasiedztwaChanged(row, col, value)),
    createEmptyMatrix: (size) => dispatch(createEmptyMatrix(size)),
  }
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(AdjacencyMatrix);
