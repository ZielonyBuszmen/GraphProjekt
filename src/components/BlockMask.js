import React from 'react';
import { Row, Col, Button } from 'reactstrap';
import { connect } from 'react-redux';
import { blockGraphInputs } from '../actions/actionCreators';

export class BlockMask extends React.Component {

  unlockMask = () => {
    this.props.unlockGraphInputs();
  };

  maskRenderer = () =>
    <div className='block'>
      <Row>
        <Col className='text-center'>
          <Button style={{'color': '#ffffff','backgroundColor': '#ff2953'}} size='lg' onClick={this.unlockMask}>Edycja</Button>
        </Col>
      </Row>
    </div>;

  render() {
    const {blockGraphInputs} = this.props;
    return blockGraphInputs ? this.maskRenderer() : null;
  }
}


const mapStateToProps = (state) => {
  return {
    blockGraphInputs: state.connectivity.blockGraphInputs,
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    unlockGraphInputs: () => dispatch(blockGraphInputs(false)),
  }
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(BlockMask);
