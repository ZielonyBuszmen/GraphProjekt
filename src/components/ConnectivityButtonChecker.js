import React from 'react';
import { Row, Col, Button, Alert} from 'reactstrap';
import { blockGraphInputs, checkGraphConnectivity } from '../actions/actionCreators';
import { connect } from 'react-redux';

class ConnectivityButtonChecker extends React.Component {


  constructor() {
    super();
    this.state = {
      checkConnectivityClicked: false
    };
  }

  checkConnectivityClick = () => {
    this.props.checkGraphConnectivity(this.props.matrix);
    this.props.blockGraphInputs(true);
  };

  messageRenderer = () => {
    const {connectivity} = this.props;

      const successMessage = <Alert  className="alert-success"><b>Graf jest silnie spójny!</b></Alert>;
      const failMessage = <Alert className="alert-danger"><b>Graf nie jest silnie spójny.</b></Alert>;

    if (connectivity.blockGraphInputs) {
      return connectivity.isGraphConnected ? successMessage : failMessage
    }
    return null;
  };

  render() {
    return (
      <div>
          {this.messageRenderer()}
        <Button style={{'color': '#ffffff','backgroundColor': '#6B0F19'}} size='lg' onClick={this.checkConnectivityClick}>Sprawdź silną spójność</Button>
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    connectivity: state.connectivity,
    matrix: state.matrix,
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    checkGraphConnectivity: (matrix) => dispatch(checkGraphConnectivity(matrix)),
    blockGraphInputs: () => dispatch(blockGraphInputs(true)),
  }
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(ConnectivityButtonChecker);