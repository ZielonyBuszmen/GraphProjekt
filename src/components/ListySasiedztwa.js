import React from 'react';
import { listySasiedztwaChanged } from '../actions/actionCreators';
import { connect } from 'react-redux';
import { immutablePush, macierzNaListe } from '../helpers';

class ListySasiedztwa extends React.Component {

  /**
   * Renderuje całą listę sąsiedztwa, wraz z inputami i wartościami
   *
   * @returns {Array}
   */
  listRenderer() {
    const {matrix} = this.props;
    const lists = macierzNaListe(matrix);
    let result = [];
    lists.map((connectedList, node) => {
      result[node] = <span>{node} - {this.inputRowRenderer(connectedList, node)}<br/></span>;
    });
    return result;
  }

  /**
   * Renderuje pojedynczy wiersz inputów
   *
   * @param list
   * @param startNode
   * @returns {Array}
   */
  inputRowRenderer = (list, startNode) => {
    let max = list.length < this.props.matrix.length ? list.length + 1 : list.length;
    let result = [];
    for (let i = 0; i < max; i++) {
      const value = list[i];
      result = immutablePush(result, <input type="text" value={value} style={{width: '35px'}} onChange={(e) => this.handleChange(e, startNode, i)}/>)
    }
    return result;
  };

  handleChange = (event, startNode, listElement) => {
    const endNode = parseInt(event.target.value);
    const lists = macierzNaListe(this.props.matrix);
    lists[startNode][listElement] = endNode;
    this.props.listySasiedztwaChanged(lists);
  };

  render() {
    return (
      <div>
        <hr/>
        Tutaj komponent z inputami z listą sąsiedztwa<br/>
        {this.listRenderer()}
        <hr/>
      </div>
    );
  }
}


const mapStateToProps = (state) => {
  return {
    matrix: state.matrix,
  };
};
const mapDispatchToProps = (dispatch) => {
  return {
    listySasiedztwaChanged: (lists) => dispatch(listySasiedztwaChanged(lists)),
  }
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(ListySasiedztwa);