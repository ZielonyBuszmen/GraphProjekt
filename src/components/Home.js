import React from 'react'
import { connect } from 'react-redux'
import MacierzSasiedztwa from './MacierzSasiedztwa';
import ListySasiedztwa from './ListySasiedztwa';
import { Row, Col, Button } from 'reactstrap';
import ConnectivityButtonChecker from './ConnectivityButtonChecker';
import '../css/style.css';
import BlockMask from './BlockMask';
import VertexQuantityInput from './VertexQuantityInput';

class Home extends React.Component {

  render() {
    const {graphSize} = this.props;

    return <div className='mt-4'>
      <p>Silnie spójna składowa jest maksymalnym podgrafem, w którym istnieją ścieżki pomiędzy każdymi dwoma wierzchołkami.
          Jeśli podgraf ten obejmuje wszystkie wierzchołki grafu, to mówimy, że dany graf skierowany jest silnie spójny (w grafach nieskierowanych każdy graf spójny jest również silnie spójny).
          Aby sprawdzić czy graf jest silnie spójny podaj liczbę wierzchołków, a później uzupełnij macierz sąsiedztwa lub listę sąsiedztw.</p>
      <br/>
      <a href='' className='btn mb-4' style={{'color': '#ffffff','backgroundColor': '#6B0F19'}}>Rozpocznij od nowa!</a>

      <Row>
        <Col>
          <BlockMask/>
          <Row>
            <Col>
              <VertexQuantityInput/>
            </Col>
          </Row>
          <Row>
            <Col md='8'>
              <b>Macierz sąsiedztwa:</b>
              <MacierzSasiedztwa size={graphSize}/>
              <hr/>
            </Col>
            <Col md='4'>
              <b>Listy sąsiedztwa:</b>
              <ListySasiedztwa/>
            </Col>
          </Row>
        </Col>
      </Row>
      <Row>
        <Col>
          <ConnectivityButtonChecker/><br/><br/><br/>
        </Col>
      </Row>
          Autorzy: Krzysztof Komar i Patrycja Madyda (Informatyka rok II, semestr IV)
    </div>;
  }

}


const mapStateToProps = (state) => {
  return {
    likes: state.likes,
    matrix: state.matrix,
    graphSize: state.graphSize,
  };
};

// const mapDispatchToProps = (dispatch) => bindActionCreators({
//   changePage: () => push('/about-us')
// }, dispatch);

const mapDispatchToProps = (dispatch) => {
  return {}
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Home);