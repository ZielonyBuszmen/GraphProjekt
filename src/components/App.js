import React from 'react';
import { Route, Link } from 'react-router-dom';
import Home from './Home';
import { Button, Row, Container, Col } from 'reactstrap';
import Header from './Header';
import { connect } from 'react-redux';
import { listySasiedztwaChanged } from '../actions/actionCreators';


class App extends React.Component {
  render() {

    const containerType = (this.props.graphSize >= 12) ? 'container-fluid' : 'container';

    return (
      <div>
        <header>
          <Header/>
        </header>

        <main>
          <div className={containerType}>
            <Row>
              <Col>
                <Home/>
              </Col>
            </Row>
          </div>
        </main>
      </div>
    );
  }
};


const mapStateToProps = (state) => {
  return {
    graphSize: state.graphSize,
  };
};
const mapDispatchToProps = (dispatch) => {
  return {}
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(App);
