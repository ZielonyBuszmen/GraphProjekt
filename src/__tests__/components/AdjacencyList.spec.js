import React from 'react';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';
import { mapDispatchToProps, mapStateToProps, AdjacencyList } from '../../components/AdjacencyList';
import { listySasiedztwaChanged } from '../../actions/actionCreators';

describe('AdjacencyList', () => {
  it('should render', () => {
    const matrix = [[1, 2], [3, 4]];
    const component = shallow(<AdjacencyList matrix={matrix}/>);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render with 4 inputs', () => {
    const matrix = [[1, 2], [3, 4]];
    const component = shallow(<AdjacencyList matrix={matrix}/>);
    expect(component.find('Input')).toHaveLength(4);
  });

  it('should render with max length', () => {
    const matrix = [[1, 0], [3, 4]];
    const component = shallow(<AdjacencyList matrix={matrix}/>);
    expect(component.find('Input')).toHaveLength(4);
  });

  it('should render with 0 inputs', () => {
    const component = shallow(<AdjacencyList matrix={[]}/>);
    expect(component.find('Input')).toHaveLength(0);
  });

  it('should change one input', () => {
    const matrix = [[1, 2], [3, 4]];
    const func = jest.fn();
    const component = shallow(<AdjacencyList matrix={matrix} listySasiedztwaChanged={func}/>);
    const inputs = component.find('Input');
    inputs.at(1).simulate('change', {target: {value: 10}});
    expect(func).toBeCalledWith([[0, 10], [0, 1]])
  });
});

describe('mapStateToProps', () => {
  it('should return object with state', () => {
    const state = {
      matrix: ['matrix'],
    };
    const result = mapStateToProps(state);
    expect(result).toEqual(state);
  });
});

describe('mapDispatchToProps', () => {
  it('should contain dispatches', () => {
    const dispatch = el => el;
    const result = mapDispatchToProps(dispatch);
    expect(result.listySasiedztwaChanged(['lists'])).toEqual(listySasiedztwaChanged(['lists']));
  });
});
