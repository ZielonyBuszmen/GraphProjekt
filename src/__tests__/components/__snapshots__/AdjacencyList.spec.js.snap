// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdjacencyList should render 1`] = `
<div
  className="macierzSasiedztwa"
>
  <table>
    <tr
      key="0"
    >
      <td>
        0
         - 
      </td>
      <td>
        <Input
          maxlength={2}
          onChange={[Function]}
          style={
            Object {
              "width": "50px",
            }
          }
          type="text"
          value={0}
        />
      </td>
      <td>
        <Input
          maxlength={2}
          onChange={[Function]}
          style={
            Object {
              "width": "50px",
            }
          }
          type="text"
          value={1}
        />
      </td>
      <br />
    </tr>
    <tr
      key="1"
    >
      <td>
        1
         - 
      </td>
      <td>
        <Input
          maxlength={2}
          onChange={[Function]}
          style={
            Object {
              "width": "50px",
            }
          }
          type="text"
          value={0}
        />
      </td>
      <td>
        <Input
          maxlength={2}
          onChange={[Function]}
          style={
            Object {
              "width": "50px",
            }
          }
          type="text"
          value={1}
        />
      </td>
      <br />
    </tr>
  </table>
</div>
`;
